pkg/operators/ebpf/tracer.go

```diff
@@ -0,0 +1,207 @@
+		return fmt.Errorf("map %q has a wrong type, expected: ringbuf or perf event array, got: %s",
+			traceMap.Name, traceMap.Type.String())
+	}
+	return nil
+}
+
+func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
+	i.logger.Debugf("populating tracer %q", varName)
+
+	parts := strings.Split(varName, typeSplitter)
+	if len(parts) != 3 {
+		return fmt.Errorf("invalid tracer info: %q", varName)
+	}
+
+	name := parts[0]
+	mapName := parts[1]
+	mapType := parts[2]
+
+	i.logger.Debugf("> name     : %q", name)
+	i.logger.Debugf("> map name : %q", mapName)
+	i.logger.Debugf("> type name: %q", mapType)
+
+	tracerConfig := i.config.Sub("tracers." + name)
+	if tracerConfig != nil {
+		if configMapName := tracerConfig.GetString("mapName"); configMapName != "" && configMapName != mapName {
+			return fmt.Errorf("mapName in tracer metadata for %q does not match metadata", varName)
+		}
+		if configMapType := tracerConfig.GetString("mapType"); configMapType != "" && configMapType != mapType {
+			return fmt.Errorf("mapName in tracer metadata for %q does not match metadata", varName)
+		}
+		i.logger.Debugf("> successfully validated with metadata")
+	}
+
+	if _, ok := i.tracers[name]; ok {
+		i.logger.Debugf("tracer %q already defined, skipping", name)
+		return nil
+	}
+
+	tracerMap, ok := i.collectionSpec.Maps[mapName]
+	if !ok {
+		return fmt.Errorf("map %q not found in eBPF object", mapName)
+	}
+
+	if err := validateTraceMap(tracerMap); err != nil {
+		return fmt.Errorf("trace map is invalid: %w", err)
+	}
+
+	var btfStruct *btf.Struct
+	if err := i.collectionSpec.Types.TypeByName(mapType, &btfStruct); err != nil {
+		return fmt.Errorf("finding struct %q in eBPF object: %w", mapType, err)
+	}
+
+	i.logger.Debugf("adding tracer %q", name)
+	i.tracers[name] = &Tracer{
+		Tracer: metadatav1.Tracer{
+			MapName:    mapName,
+			StructName: btfStruct.Name,
+		},
+	}
+
+	err := i.populateStructDirect(btfStruct)
+	if err != nil {
+		return fmt.Errorf("populating struct %q for tracer %q: %w", btfStruct.Name, name, err)
+	}
+
+	return nil
+}
+
+func (t *Tracer) receiveEvents() error {
+	switch t.mapType {
+	case ebpf.RingBuf:
+		return t.receiveEventsFromRingReader()
+	case ebpf.PerfEventArray:
+		return t.receiveEventsFromPerfReader()
+	default:
+		return fmt.Errorf("invalid map type")
+	}
+}
+
+func (t *Tracer) receiveEventsFromRingReader() error {
+	for {
+		rec, err := t.ringbufReader.Read()
+		if err != nil {
+			return err
+		}
+		data := t.ds.NewData()
+		err = t.accessor.Set(data, rec.RawSample)
+		if err != nil {
+			log.Printf("error setting buffer: %v", err)
+			t.ds.Release(data)
+			continue
+		}
+		err = t.ds.EmitAndRelease(data)
+		if err != nil {
+			log.Printf("error emitting data: %v", err)
+		}
+	}
+}
+
+func (t *Tracer) receiveEventsFromPerfReader() error {
+	for {
+		rec, err := t.perfReader.Read()
+		if err != nil {
+			return err
+		}
+		data := t.ds.NewData()
+		err = t.accessor.Set(data, rec.RawSample)
```

Comment:
The `Read` function can return some trailing bytes:

> // Records can contain between 0 and 7 bytes of trailing garbage from the ring
> // depending on the input sample's length.

This is causing an error when perf event array is used. Consider this gadget as an example:

```c
#include <vmlinux.h>
#include <bpf/bpf_helpers.h>
#include <gadget/buffer.h>
#include <gadget/macros.h>

struct event {
	__u32 pid;
	__u32 uid;
};

struct {
	__uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
	__uint(key_size, sizeof(u32));
	__uint(value_size, sizeof(u32));
} events SEC(".maps");

// [Optional] Define a tracer
GADGET_TRACER(open, events, event);

SEC("tracepoint/syscalls/sys_enter_openat")
int enter_openat(struct syscall_trace_enter *ctx)
{
	struct event event = {};

	event.pid = bpf_get_current_pid_tgid() >> 32;
	event.uid = bpf_get_current_uid_gid();

 	bpf_perf_event_output(ctx, &events, BPF_F_CURRENT_CPU, &event, sizeof(event));

	return 0;
}

char LICENSE[] SEC("license") = "GPL";
``` 

```bash 
$ sudo -E ig image build . -t foo

$ sudo -E ig run-experimental foo
INFO[0000] Experimental features enabled
PID                                                                             UID
INFO[0000] error setting buffer: invalid size, expected 8, got 12
INFO[0000] error setting buffer: invalid size, expected 8, got 12
```

It can be fixed by using something like:

```diff
diff --git pkg/operators/ebpf/tracer.go pkg/operators/ebpf/tracer.go
index 26c2740f0..8445af067 100644
--- pkg/operators/ebpf/tracer.go
+++ pkg/operators/ebpf/tracer.go
@@ -38,6 +38,7 @@ type Tracer struct {
        accessor datasource.FieldAccessor

        mapType       ebpf.MapType
+       eventSize     uint32 // needed to trim trailing bytes when reading for perf event array
        ringbufReader *ringbuf.Reader
        perfReader    *perf.Reader
 }
@@ -102,6 +103,7 @@ func (i *ebpfInstance) populateTracer(t btf.Type, varName string) error {
                        MapName:    mapName,
                        StructName: btfStruct.Name,
                },
+               eventSize: btfStruct.Size,
        }

        err := i.populateStructDirect(btfStruct)
@@ -150,7 +152,7 @@ func (t *Tracer) receiveEventsFromPerfReader() error {
                        return err
                }
                data := t.ds.NewData()
-               err = t.accessor.Set(data, rec.RawSample)
+               err = t.accessor.Set(data, rec.RawSample[:t.eventSize])
                if err != nil {
                        log.Printf("error setting buffer: %v", err)
                        t.ds.Release(data)
```